internal/beater/server.go

```diff
@@ -222,11 +222,10 @@ func (s server) run(ctx context.Context) error {
 	})
 	g.Go(func() error {
 		<-ctx.Done()
-		// httpServer should stop before grpcServer to avoid a panic caused by placing a new connection into
-		// a closed grpc connection channel during shutdown.
-		// See https://github.com/elastic/gmux/issues/13
-		s.httpServer.stop()
 		s.grpcServer.GracefulStop()
+		stopctx, cancel := context.WithTimeout(context.Background(), s.cfg.ShutdownTimeout)
+		defer cancel()
+		s.httpServer.stop(stopctx)
```

Comment:
why not just propagate `ctx` that already exists here, it should be already canceled

Reply:
That's the ctx that get canceled on signal receive. We are supposed to wait for a while to gracefully shutdown and drain connections. see https://github.com/elastic/apm-server/blob/6ad5821fcf930ce5ca8f3aafc02609752c3b24e1/internal/beater/beater.go#L165

Reply:
Oh I didn't notice that we are using `ShutdownTimeout` here, though we just propagate an independent cancel context. Looks good then!